trlx/trainer/nemo_ppo_trainer.py

```diff
@@ -0,0 +1,377 @@
+
+        megatron_cfg.trlx = config.to_dict()
+
+        megatron_cfg.model.global_batch_size = config.train.batch_size
+        megatron_cfg.model.seed = config.train.seed
+
+        world_size = megatron_cfg.trainer.num_nodes * megatron_cfg.trainer.devices
+        dp_world = world_size // (
+            megatron_cfg.model.tensor_model_parallel_size * megatron_cfg.model.pipeline_model_parallel_size
+        )
+        rank_batch_size = megatron_cfg.model.global_batch_size // dp_world
+
+        if (self.ppo_config.chunk_size % rank_batch_size) != 0:
+            self.ppo_config.chunk_size = rank_batch_size * ceil(self.ppo_config.chunk_size / rank_batch_size)
+            print("Rounding chunk size to", self.ppo_config.chunk_size)
+
+        train_samples = dp_world * self.ppo_config.num_rollouts * self.ppo_config.ppo_epochs
+        # Disable validation within nemo, run it ourselves
+        self.limit_val_batches = megatron_cfg.trainer.limit_val_batches
+        self.val_check_interval = config.train.eval_interval
+
+        megatron_cfg.trainer.limit_val_batches = 0.0
+        megatron_cfg.trainer.val_check_interval = None
+        self.train_samples = train_samples
+        megatron_cfg.trainer.max_steps = config.train.epochs * (train_samples // megatron_cfg.model.global_batch_size)
+
+        self.trainer = megatron_trainer(megatron_cfg)
+
+        self.model = PPOGPT(
+            ppo_config=self.ppo_config,
+            cfg=megatron_cfg.model,
+            trainer=self.trainer,
+            metric_fn=self.metric_fn,
+            stop_sequences=self.stop_sequences,
+            num_layers_unfrozen=config.model.num_layers_unfrozen,
+        )
+        self.megatron_cfg = megatron_cfg
+
+        if pretrained_model is not None:
+            self.model.load_from_pretrained(pretrained_model)
+
+        self.batch_size = megatron_cfg.model.global_batch_size
+        self.tokenizer = self.model.tokenizer.tokenizer
+        self.tokenizer.truncation_side = config.tokenizer.truncation_side
+        if self.tokenizer.pad_token is None:
+            self.tokenizer.pad_token_id = self.tokenizer.eos_token_id
+            self.tokenizer.pad_token = self.tokenizer.eos_token
+
+        self.max_length = megatron_cfg.model.encoder_seq_length
+
+        if self.ppo_config.target is not None:
+            self.kl_ctl = AdaptiveKLController(
+                self.ppo_config.init_kl_coef, self.ppo_config.target, self.ppo_config.horizon
+            )
+        else:
+            self.kl_ctl = FixedKLController(self.ppo_config.init_kl_coef)
+
+    def add_prompt_pipeline(self, pipeline: PromptPipeline):
+        self.prompt_pipeline = pipeline
+
+    def make_experience(self, prompt_iterator: Iterator, num_rollouts: int = 1024, dp_world: int = 1):
+        ppo_rl_elements = []
+
+        device = torch.device("cuda")
+
+        if torch.distributed.get_rank() == 0:
+            tbar = tqdm(total=num_rollouts * dp_world, desc="Generating experience")
```

Comment:
Nit:
```suggestion
            tbar = tqdm(total=num_rollouts * dp_world, desc="Generating experience", disable=torch.distributed.get_rank())
```
This should remove the need for checks in further `update`/`close` calls

Reply:
Cleaned up the tbar handling a bit